File: src/find_available_listings.py
import time
import sys, json
from first_fit import first_fit_decreasing

def find_available_listings():
    print("Starting the first fit algorithm...", file=sys.stderr)

    # read the and convert the posted json
    vehicles_data = json.load(sys.stdin)

    items = []

    for item in vehicles_data:
        for i in range(item["quantity"]):
            items.append(item["length"])

    
    # read the listings json
    with open('listings.json', 'r') as f:
        listings_data = json.load(f)
    
    available_listings = []

    for listing in listings_data:
        bin_capacity = listing["length"]
        bins = first_fit_decreasing(items, bin_capacity)
        if len(bins) <= listing["width"] / 10:
            available_listings.append(listing)

    # Print the result as JSON
    json.dump(available_listings, sys.stdout)

if __name__ == "__main__":
    find_available_listings()

File: src/findAvailableListings.js
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { findCheapestCombination } from './findCheapestCombination.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

function findAvailableListings(vehicles_data) {
    const items = [];

    // Convert vehicles data to array of vehicle lengths
    for (const item of vehicles_data) {
        for (let i = 0; i < item.quantity; i++) {
            items.push(item.length);
        }
    }

    // Read the listings json
    const listingsPath = join(__dirname, '..', 'listings.json');
    const listings_data = JSON.parse(
        readFileSync(listingsPath, 'utf-8')
    );

    // Group listings by location_id
    const locationMap = new Map();
    
    for (const listing of listings_data) {
        if (!locationMap.has(listing.location_id)) {
            locationMap.set(listing.location_id, []);
        }
        locationMap.get(listing.location_id).push(listing);
    }

    const results = [];

    // For each location, find if vehicles can fit and find cheapest combination
    for (const [location_id, listings] of locationMap) {
        const cheapestCombination = findCheapestCombination(listings, items);
        
        if (cheapestCombination) {
            results.push({
                location_id,
                listing_ids: cheapestCombination.listing_ids,
                total_price_in_cents: cheapestCombination.total_price
            });
        }
    }

    // Sort by total price ascending
    results.sort((a, b) => a.total_price_in_cents - b.total_price_in_cents);

    return results;
}

export { findAvailableListings };

